airflow/providers/hashicorp/secrets/vault.py

```diff
@@ -0,0 +1,196 @@
+    :type secret_id: str
+    :param gcp_key_path: Path to GCP Credential JSON file (for ``gcp`` auth_type)
+    :type gcp_key_path: str
+    :param gcp_scopes: Comma-separated string containing GCP scopes (for ``gcp`` auth_type)
+    :type gcp_scopes: str
+    """
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        path: str,
+        url: Optional[str] = None,
+        auth_type: str = 'token',
+        mount_point: str = 'secret',
+        token: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        role_id: Optional[str] = None,
+        secret_id: Optional[str] = None,
+        gcp_key_path: Optional[str] = None,
+        gcp_scopes: Optional[str] = None,
+        **kwargs
+    ):
+        super().__init__()
+        self.path = path.rstrip('/')
+        self.url = url
+        self.auth_type = auth_type
+        self.kwargs = kwargs
+        self.token = token
+        self.username = username
+        self.password = password
+        self.role_id = role_id
+        self.secret_id = secret_id
+        self.mount_point = mount_point
+        self.gcp_key_path = gcp_key_path
+        self.gcp_scopes = gcp_scopes
+        self._client: Optional[hvac.Client] = None
+
+    def get_client(self) -> hvac.Client:
+        """
+        Return an authenticated Hashicorp Vault client
+        """
+        if not self._client:
+            self._client = hvac.Client(url=self.url, **self.kwargs)
+            if self.auth_type == "token":
+                self._client.token = self.token
+            elif self.auth_type == "ldap":
+                self._client.auth.ldap.login(
+                    username=self.username, password=self.password)
+            elif self.auth_type == "userpass":
+                self._client.auth_userpass(username=self.username, password=self.password)
+            elif self.auth_type == "approle":
+                self._client.auth_approle(role_id=self.role_id, secret_id=self.secret_id)
+            elif self.auth_type == "github":
+                self._client.auth.github.login(token=self.token)
+            elif self.auth_type == "gcp":
+                credentials = self._get_gcp_credentials()
+                self._client.auth.gcp.configure(credentials=credentials)
+            else:
+                raise AirflowException(f"Authentication type '{self.auth_type}' not supported")
+
+            if self._client.is_authenticated():
+                return self._client
+            else:
+                raise VaultError("Vault Authentication Error!")
+        else:
+            return self._client
+
+    def get_conn_uri(self, conn_id: str) -> Optional[str]:
+        """
+        Get secret value from Vault. Store the secret in the form of URI
+
+        :param conn_id: connection id
+        :type conn_id: str
+        """
+        connection_id = f"{CONN_ENV_PREFIX}{conn_id}".upper()
```

Comment:
Why do we have this if we have a path already? This seems a bit odd -- for instance I'd expect tit to be something like `airflow/connections/mysql_default`.

Reply:
This was done to keep it consistent with how AWS SSM has implemented i.e. store keys as "AIRFLOW_CONN_{CONN_ID}" format.

But I agree it is redundant and does not make sense.  I have updated it in https://github.com/apache/airflow/pull/7741/commits/d59b45aace618e5d373eb02f421f53bc4352062b

Reply:
@dstandish what do you think of making the same change to the AWS SSM provider (before we include it in a replace)

Reply:
I had a talk with @dstandish and he agrees we should update AWS SSM. He is busy today, so I might make the change, if I don't get to it today, he might do it tomorrow.

Reply:
the thinking was just to keep things uniform between env vars and creds store so you could copy paste and not think about it or have to change the keys in text editor. 

but this obvs has benefit of readability and simplicity, when looking at the creds store in isolation alone.

one little thing... if @potiuk's idea materializes -- generic arbitrary secret...  would a prefix be helpful to distinguish between types?  or would it just pull with same convention / prefix, get secret is raw value, get connection pulls same key but attempts to parse as connection....

Reply:
I think in most cases 1 prefix might be fine. Just `/airflow` might do. When we implement generic secret, we can provide user a way to override/specify prefix/entire path from the operator/hook/xcom etc

Reply:
no i agree but what i mean is like it could be like `AIRFLOW_SECRET_BLAH` and `AIRFLOW_CONN_BLAH` in theory.... and allow it to work on env vars too... and same prefix....

just a thought, a hypothetical

Reply:
With the example as @kaxil has it in the docs it's already "namespaced" as `airflow/connections/`

Reply:
but then what if you add in `get_secret`.  do you have to modify backends to have two prefixes -- one for connections, and one for secrets?

Reply:
Ohhh good point. Let's address that in the future I think?

Or we could change the `prefix` argument to `connections_prefix` now to future proof it a bit?

Reply:
> Or we could change the prefix argument to connections_prefix now to future proof it a bit?

I think that seems like a good idea, unless you want  connection to merely be a special case of secret, i.e. in same "namespace".

so e.g. ...
if ssm param is `/airflow/smtp_default=my://uri` and prefix is `/airflow`, then `get_secret` pulls the raw value `my://uri` and get_connections pulls the same raw value _and_ tries to parse the URI and load as connection....

if i had to choose, i would probably say, do as you've suggested -- put them on different prefixes, and change to `connections_prefix` now...

Reply:
Do we need to do that though? I would suggest changing the `get_conn_uri` method on VaultSecrets or AwsSsmBackend to a generic `get_secret` method as follows:

```python
    def get_secret(
        self, secret_key: str, path: Optional[str] = None, mount_point: Optional[str] = None
    ) -> Optional[str]:
        """
        Get secret value from Vault

        :param secret_key: Key of the Secret to retrieve
        :type secret_key: str
        :param path: Path of the secret to read
        :type path: str
        :param mount_point: The "path" the secret engine was mounted on
        :type mount_point: str
        """
        path = path if path else self.path
        mount_point = mount_point if mount_point else self.mount_point

        self.log.debug("Path: %s", path)
        self.log.debug("Mount Point: %s", mount_point)
        self.log.debug("Retrieving the secret for the key: %s", secret_key)

        if self.kv_engine_version == 1:
            response = self.client.secrets.kv.v1.read_secret(
                path=self.path, mount_point=self.mount_point
            )
        else:
            response = self.client.secrets.kv.v2.read_secret_version(
                path=self.path, mount_point=self.mount_point)

        if not response:
            return None

        return_data = response["data"] if self.kv_engine_version == 1 else response["data"]["data"]
        return return_data.get(secret_key)
```

This method would allow overwriting the **path** and **mount_point** and hence we can use this method in Operator/xcom

I think we can figure out later when we actually implement that generic method.

Reply:
I think mount_point is a Vault-specific concept, but path is probably generic enough a concept.

Reply:
The main reason for having a "dedicated" way of getting connections is so that we could have the way of listing the connections - checking if an individual connection exists could just try and read the secret and parse it, but if we want to support list then we need a way to tell connections from other secrets.

Reply:
**Updated**